### PR TITLE
cleanup cache directory after umount

### LINKED ISSUE #328: [FEATURE] Cleanup cached blocks when PV is not used anymore
Maybe we need a config?

### PR DESCRIPTION
**Pull request by zwwhdls**
fix https://github.com/juicedata/juicefs-csi-driver/issues/328

### CODE REVIEW

File: pkg/juicefs/mount/process_mount.go [hunk 1]
```diff
@@ -219,6 +221,23 @@ func (p *ProcessMount) AddRefOfMount(target string, podName string) error {
...
 }
 
+func (p *ProcessMount) CleanCache(id string, volumeId string, cacheDirs []string) error {
+	for _, cacheDir := range cacheDirs {
+		// clean up raw dir under cache dir
+		rawPath := filepath.Join(cacheDir, id, "raw")
+		if existed, err := k8sMount.PathExists(rawPath); err != nil {
+			return status.Errorf(codes.Internal, "Could not check raw path %q exists: %v", rawPath, err)
+		} else if existed {
+			stdoutStderr, err := p.Exec.Command("rm", "-rf", rawPath).CombinedOutput()
```
Comment (davies) [inline pkg/juicefs/mount/process_mount.go]: nit: os.RemoveAll()

File: pkg/juicefs/mount/pod_mount.go [hunk 2]
```diff
@@ -348,6 +350,39 @@ func (p *PodMount) AddRefOfMount(target string, podName string) error {
...
+		UUID:      id,
+	}
+	r := builder.NewBuilder(jfsSetting)
+	job := r.NewJobForCleanCache()
+	klog.V(6).Infof("Clean cache job: %v", job)
+	_, err := p.K8sClient.GetJob(job.Name, job.Namespace)
+	if err != nil && k8serrors.IsNotFound(err) {
+		klog.V(5).Infof("CleanCache: create job %s", job.Name)
+		_, err = p.K8sClient.CreateJob(job)
+		if err != nil {
```
Comment (davies) [inline pkg/juicefs/mount/pod_mount.go]: will be checked later

File: pkg/juicefs/mount/builder/common.go [hunk 3]
```diff
@@ -160,3 +162,50 @@ func (r *Builder) getVolumeMounts() []corev1.VolumeMount {
...
+			Labels: map[string]string{
+				config.PodTypeKey: config.PodTypeValue,
+			},
+			Annotations: make(map[string]string),
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{{
+				Name:         "jfs-cache-clean",
+				Image:        config.MountImage,
+				Command:      []string{"sh", "-c", "rm -rf /var/jfsCache/*/chunks/*"},
```
Comment (davies) [inline pkg/juicefs/mount/builder/common.go]: The last `*` is not needed

Comment (zwwhdls) [reply pkg/juicefs/mount/builder/common.go]: remove dir `chunks` is ok?

File: pkg/juicefs/juicefs.go [hunk 4]
```diff
@@ -222,14 +222,13 @@ func (j *juicefs) getSettings(volumeID string, target string, secrets, volCtx ma
...
 		}
-		// Default use redis:// scheme
-		if !strings.Contains(source, "://") {
-			source = "redis://" + source
-		}
-		jfsSetting.Source = source
 		if config.FormatInPod {
 			jfsSetting.FormatCmd = res
 		}
+		jfsSetting.UUID, err = j.GetJfsVolUUID(jfsSetting.Source)
```
Comment (davies) [inline pkg/juicefs/juicefs.go]: If cleanCache is not set, we don't need UUID

File: pkg/controller/pod_driver.go [hunk 5]
```diff
@@ -544,3 +547,45 @@ func (p *PodDriver) removeFinalizer(pod *corev1.Pod) error {
...
 }
+
+func (p PodDriver) CleanUpCache(pod *corev1.Pod) {
+	if pod.Annotations[config.CleanCache] != "true" {
+		return
+	}
+	uuid := pod.Annotations[config.JuiceFSUUID]
+	uniqueId := pod.Annotations[config.UniqueId]
+	if uuid == "" && uniqueId == "" {
+		// no necessary info, return
```
Comment (davies) [inline pkg/controller/pod_driver.go:558]: an warning?

File: pkg/juicefs/juicefs.go [hunk 6]
```diff
@@ -262,45 +287,82 @@ func (j juicefs) getUniqueId(volumeId string) (string, error) {
...
+					return
+				}
+				delete(j.UUIDMaps, uniqueId)
+				delete(j.CacheDirMaps, uniqueId)
 
-	// cleanup target path
-	if err := j.JfsCleanupMountPoint(mountPath); err != nil {
-		klog.V(5).Infof("Clean mount point error: %v", err)
+				klog.V(5).Infof("Cleanup cache of volume %s in node %s", uniqueId, config.NodeName)
+				if err = j.processMount.CleanCache(uuid, uniqueId, cacheDirs); err != nil {
```
Comment (davies) [inline pkg/juicefs/juicefs.go]: this could be slow, and block `j`

**Review by davies — APPROVED** [review]

**Pull request merged**